src/db/migration/mod.rs

```diff
@@ -0,0 +1,129 @@
+    ensure_params_downloaded, set_proofs_parameter_cache_dir_env,
+};
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tracing::info;
+
+pub const LATEST_DB_VERSION: DBVersion = DBVersion::V11;
+
+/// Database version for each forest version which supports db migration
+#[derive(Debug, Eq, PartialEq)]
+pub enum DBVersion {
+    V0, // Default DBVersion for any unknown db
+    V11,
+}
+
+/// Check to verify database migrations
+async fn migration_check(config: &Config, existing_chain_data_root: &Path) -> anyhow::Result<()> {
+    info!(
+        "Running database migration checks for: {}",
+        existing_chain_data_root.display()
+    );
+    // Set proof param dir env path, required for running validations
+    if cns::FETCH_PARAMS {
+        set_proofs_parameter_cache_dir_env(&config.client.data_dir);
+    }
+    ensure_params_downloaded().await?;
+
+    // Open db
+    let db = Arc::new(open_proxy_db(
+        db_root(existing_chain_data_root),
+        config.db_config().clone(),
+    )?);
+    let genesis = read_genesis_header(None, config.chain.genesis_bytes(), &db).await?;
+    let chain_store = Arc::new(ChainStore::new(
+        db,
+        Arc::clone(&config.chain),
+        genesis,
+        existing_chain_data_root,
+    )?);
+    let state_manager = Arc::new(StateManager::new(chain_store, Arc::clone(&config.chain))?);
+
+    let ts = state_manager.chain_store().heaviest_tipset();
+    let height = ts.epoch();
+    assert!(height.is_positive());
+    // re-compute 100 tipsets only
+    state_manager.validate_range((height - 100)..=height)?;
+    Ok(())
+}
+
+/// Migrate database to latest version
+pub async fn migrate_db(
+    config: &Config,
+    db_path: PathBuf,
+    target_version: DBVersion,
+) -> anyhow::Result<()> {
+    info!("Running database migrations...");
+    // Get DBVersion from existing db path
+    let mut current_version = get_db_version(&db_path);
+    // Run pre-migration checks, which includes:
+    // - re-compute 100 tipsets
+    migration_check(config, &db_path).await?;
+
+    // Iterate over all DBVersion's until database is migrated to lastest version
+    while current_version != target_version {
+        let next_version = match current_version {
+            DBVersion::V0 => DBVersion::V11,
+            _ => break,
+        };
+        // Execute the migration steps for itermediate version
+        migrate(&db_path, &next_version)?;
+        current_version = next_version;
+    }
+    // Run post-migration checks, which includes:
+    // - re-compute 100 tipsets
+    migration_check(config, &db_path).await?;
+
+    // Rename db to latest versioned db
+    fs::rename(db_path.as_path(), chain_path(config))?;
+
+    info!("Database Successfully Migrated to {:?}", target_version);
+    Ok(())
+}
+
+// TODO: Add Steps required for new migration
+/// Migrate to an intermediate db version
+fn migrate(_existing_db_path: &Path, next_version: &DBVersion) -> anyhow::Result<()> {
```

Comment:
It looks like you want the migrations to happen _inside_ the database folder. That's not safe. If something goes wrong, we want to revert back to the last known good state.

Reply:
I am currently using `temp_dir` for intermediate db migrations, if something goes wrong we can revert back to last good state easily